src/posts/aphrodite-inline-css.md

```diff
@@ -0,0 +1,338 @@
+    }
+});
+
+const Component2 = () => <span className={css(styles2.red)}>2</span>;
+
+const styles2 = StyleSheet.create({
+    red: {
+        backgroundColor: '#cc1100'
+    }
+});
+```
+
+It would be surprising behaviour if either one of those ended up using the other's style, so we need some way to differentiate between them. We initially solved this by appending a random number to the class name. This works great… unless you want server-side rendering to work well.
+
+## Deterministic rendering and data-react-checksum
+
+When you render a component server-side via ReactDOMServer.renderToString, out outputs HTML that might look something like this:
+
+```
+<div id='container'
+    data-reactid=".1x73ayqlreo"
+    data-react-checksum="-1952287665">...</div>
+```
+
+From working with React keys, you've likely seen `data-reactid` before, but 
+what's this `data-react-checksum` business? 
+
+When your re-render your React component on the client-side to bind all your 
+events and run the lifecycle starting with `componentDidMount`, 
+`ReactDOM.render`
+will construct a virtual DOM using your component. It then takes a checksum of 
+that component tree and compares it with the one emitted by 
+`ReactDOMServer.renderToString`. If it's identical it will re-use the 
+server-rendered DOM tree and avoid needing to reflow the page or do any style 
+recalculations. If it's different, it will replace the entire DOM tree.
+
+This is not so great for performance, so we'd like to avoid that.
+
+If we generate random numbers that wind up in the HTML output of our component, then the checksum certainly won't match. So we can't use random numbers, but we still need something to avoid name collisions.
+
+We could use a simple global counter, but when would we reset that counter? Each time we render? If we do that, then we might still get name collisions between classes generated by two separate render calls.
+
+What we really want is a 1:1 mapping between the class name and the style declarations associated with that class name. Sounds like a job for hashing!
+
+We take a hash of the JSON stringified style declaration and use that as a suffix. On recommendation from our Dean of Infrastructure, Craig Silverstein, we used [MurmurHash](https://en.wikipedia.org/wiki/MurmurHash).
+
+It is possible that two components will each declare their own copy of styles with the same name and the same contents, like so:
+
+```
+const Component1 = () => <span className={css(styles1.red)}>1</span>
+
+const styles1 = StyleSheet.create({
+    red: {
+        backgroundColor: 'red'
+    }
+});
+
+const Component2 = () => <span className={css(styles2.red)}>2</span>
+
+const styles2 = StyleSheet.create({
+    red: {
+        backgroundColor: 'red'
+    }
+});
+```
+
+In this case, we get a class name collision on `red_im3wl1`, but that's okay, 
+because they reference the exact same properties and values, so there's no 
+surprising behaviour!
+
+## Deterministic precedence
+
```

Comment:
Maybe add a line here like "One of the things we really wanted to avoid was the unexpected and spooky-action-at-a-distance part of CSS that is precedence resolution. In a stylesheet, where there are two styles defined with equal specificity, the one that comes *later* in the stylesheet has precedence. In a hand-tuned stylesheet that can be confusing, but in an auto-generated one it can be disastrous!"